realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -1865,3 +1865,40 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty
         }
     } CATCH_STD()
 }
+
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeDistinct
+    (JNIEnv *env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes) {
+
+    JniLongArray arr(env, columnIndexes);
+    jlong column_idx = arr[arr.len() - 1];
+    TableRef table_ref = getTableForLinkQuery(nativeQueryPtr, arr);
+
+    Query* pQuery = Q(nativeQueryPtr);
+
+    try {
+
+        if (!QUERY_VALID(env, pQuery) || !TBL_AND_COL_INDEX_VALID(env, table_ref.get(), column_idx))
+            return;
+
+        if (!table_ref->has_search_index(S(column_idx))) {
+            ThrowException(env, UnsupportedOperation, "The field must be indexed before distinct() can be used.");
+        }
+
+        // Field queries
+        int col_type = table_ref->get_column_type(S(column_idx));
+        switch (col_type) {
+            case type_Bool:
+            case type_Int:
+            case type_DateTime:
+            case type_String:
+                try {
+                    pQuery->and_query(Query(table_ref->get_distinct_view(S(column_idx)).get_query()));
```

Comment:
I am not sure this will work correctly, as the distinct_view quite possibly still have the same backing query. See my comment on the Java API. I think this just needs to run the query then restrict it to the distinct view